miniml/compiler/compile.scm

```diff
@@ -1793,6 +1823,98 @@
+           env exit-names exit-infos))
+      (max-arity
+       (fold (lambda (info max-arity) (max (exit-info-get-arity info) max-arity))
+           0 exit-infos))
+    )
+    ; Each exit needs an "argument space" of size exit-arity reserved on the stack.
+    ; Jumping to one of the exits leaves the scope were those exits are available,
+    ; so we will jump to at most one of these exit.
+    ; We reserve overlapping space for their arguments, of size max-arity.
+    (bytecode-put-u32-le ATOM0)
+    (for-each (lambda (i)
+        (bytecode-put-u32-le PUSH)
+      ) (range 0 max-arity))
+    (compile-expr body-env (+ stacksize max-arity) body)
+    (bytecode-put-u32-le POP)
+    (bytecode-put-u32-le max-arity)
+    (bytecode-put-u32-le BRANCH)
+    (bytecode-emit-labref endlab)
+    (for-each (match-lambda* ((($ <exit-info> exit-label exit-arity exit-env exit-stacksize) exit-body)
+         (bytecode-emit-label exit-label)
+         (compile-expr exit-env exit-stacksize exit-body)
+         (bytecode-put-u32-le POP)
+         (bytecode-put-u32-le exit-arity)
+         (assert (equal? exit-arity (- exit-stacksize stacksize)))
+         (bytecode-put-u32-le BRANCH)
+         (bytecode-emit-labref endlab)
+       )) exit-infos exit-bodies)
+    (bytecode-emit-label endlab)
+  ))
+
+(define (compute-exit-info env stacksize exit-label exit-vars)
+  (let* (
+    (arity
+     (length exit-vars))
+    (exit-env
+     (fold (lambda (var i env)
+             (stack-var env var (+ stacksize i))
+       ) env exit-vars (range 0 arity)))
+    (exit-stacksize (+ stacksize arity))
+    )
+    (make-exit-info exit-label arity exit-env exit-stacksize)))
+
+(define (compile-exit env stacksize exit args)
+  (match-let*
+     ((($ <exit-info> exit-label arity exit-env exit-stacksize) (compenv-get-exit env exit))
+      ; this exit has a reserved "argument space" of the stack of size arity,
+      ; placed immediately *before* exit-stacksize, to which the exit arguments
+      ; must be assigned.
+      (argspace-start (- exit-stacksize arity)))
+  (assert (equal? arity (length args)))
+  ; fill the argument space
+  ; (this should go first, before we change the traps or stack, as it runs arbitrary code)
+  (for-each (lambda (arg i)
+       (compile-expr env stacksize arg)
+       (bytecode-put-u32-le ASSIGN)
+       (bytecode-put-u32-le (stack-relative stacksize (+ argspace-start i)))
+     ) args (range 0 arity))
+  ; pop all the exception traps pushed between the exit stack and the current stack
+  (let* (
+    (exit-traps (compenv-get-traps exit-env))
+    (stacksize
+     (let pop-handlers ((stacksize stacksize) (traps (compenv-get-traps env)))
+       (let ((same-handlers
+              (if (null? exit-traps) (null? traps) (equal? (car exit-traps) (car traps)))))
+       (if (not same-handlers)
+           (begin
+             (bytecode-put-u32-le POPTRAP)
+             (pop-handlers (- stacksize 4) (cdr traps)))
```

Comment:
It seems to me that this will put a certain number of `POPTRAPS`, followed by a `POP`, but this might not be the correct behavior: it seems to me that `POPTRAP` will only work if the exception handler is at the top of the stack, so we would need alternating `POP`s and `POPTRAP`s.

I have not tested it yet, but I think something like the following will cause an incorrect exception handler to be popped:
```ocaml
let () =
  print_int (
    let%exit ex n = n + 2 in
    try
      let dummy = 42 in
      [%exit] ex 3
    with Foo n -> print_int n
  )
```

Reply:
Oh, I forgot that POPTRAP should only happen when the trap is at the top of the stack. This explains why I am not currently using the stack heights I record for trap handlers... Will fix. Thanks!

Reply:
Amusingly, your example does pass currently. POPTRAP only reads the trap-handler slots to reset the `trapsp` pointer to its previous value; in your example, `trapsp` gets corrupted, but this does not matter as it is not later used. I could observe a Real Crash by having another exception handler around it:

```
let () =
  print_int (
    try 
      let%exit ex n = raise (Foo (n + 2)) in
      try
        let v = 1 in
        [%exit] ex v
      with Foo n -> print_int (2 * n)
    with Foo n -> print_int n
  )
```

Reply:
I have now implemented proper stack-walking (but I haven't fixed the frontend parsing yet).